proof/invariant-abstract/AARCH64/ArchIpc_AI.thy

```diff
@@ -399,109 +411,128 @@ lemma transfer_caps_loop_valid_vspace_objs[wp, Ipc_AI_assms]:
+  for pred_tcb[wp, Ipc_AI_2_assms]: "pred_tcb_at proj P t"
 crunch  make_arch_fault_msg
-  for cap_to[wp, Ipc_AI_assms]: "ex_nonz_cap_to p"
+  for cap_to[wp, Ipc_AI_2_assms]: "ex_nonz_cap_to p"
 
 crunch  make_arch_fault_msg
-  for v_ker_map[wp, Ipc_AI_assms]: "valid_kernel_mappings"
+  for v_ker_map[wp, Ipc_AI_2_assms]: "valid_kernel_mappings"
  (simp: valid_kernel_mappings_def)
 
 crunch  make_arch_fault_msg
-  for obj_at[wp, Ipc_AI_assms]: "\<lambda>s. P (obj_at P' pd s)"
+  for obj_at[wp, Ipc_AI_2_assms]: "\<lambda>s. P (obj_at P' pd s)"
   (wp: as_user_inv getRestartPC_inv mapM_wp'  simp: getRegister_def)
 
 lemma dmo_addressTranslateS1_valid_machine_state[wp]:
   "do_machine_op (addressTranslateS1 addr) \<lbrace> valid_machine_state \<rbrace>"
   by (wpsimp wp: dmo_valid_machine_state)
 
 crunch make_arch_fault_msg
-  for vms[wp, Ipc_AI_assms]: valid_machine_state
+  for vms[wp, Ipc_AI_2_assms]: valid_machine_state
   (wp: as_user_inv getRestartPC_inv mapM_wp'  simp: getRegister_def ignore: do_machine_op)
 
 lemma dmo_addressTranslateS1_valid_irq_states[wp]:
   "do_machine_op (addressTranslateS1 addr) \<lbrace> valid_irq_states \<rbrace>"
   by (wpsimp wp: dmo_valid_irq_states)
 
 crunch make_arch_fault_msg
-  for valid_irq_states[wp, Ipc_AI_assms]: "valid_irq_states"
+  for valid_irq_states[wp, Ipc_AI_2_assms]: "valid_irq_states"
   (wp: as_user_inv getRestartPC_inv mapM_wp'  simp: getRegister_def ignore: do_machine_op)
 
 lemma dmo_addressTranslateS1_cap_refs_respects_device_region[wp]:
   "do_machine_op (addressTranslateS1 addr) \<lbrace> cap_refs_respects_device_region \<rbrace>"
   by (wpsimp wp: cap_refs_respects_device_region_dmo)
 
 crunch make_arch_fault_msg
-  for cap_refs_respects_device_region[wp, Ipc_AI_assms]: "cap_refs_respects_device_region"
+  for cap_refs_respects_device_region[wp, Ipc_AI_2_assms]: "cap_refs_respects_device_region"
   (wp: as_user_inv getRestartPC_inv mapM_wp'  simp: getRegister_def ignore: do_machine_op)
 
+lemma setup_caller_cap_aobj_at:
+  "arch_obj_pred P' \<Longrightarrow>
+  \<lbrace>\<lambda>s. P (obj_at P' pd s)\<rbrace> setup_caller_cap st rt grant \<lbrace>\<lambda>r s. P (obj_at P' pd s)\<rbrace>"
+  unfolding setup_caller_cap_def
+  by (wpsimp wp: cap_insert_aobj_at sts.aobj_at)
+
+lemma setup_caller_cap_valid_arch[Ipc_AI_2_assms, wp]:
+  "setup_caller_cap st rt grant \<lbrace>valid_arch_state\<rbrace>"
+  by (wpsimp wp: valid_arch_state_lift_aobj_at_no_caps[rotated -1] setup_caller_cap_aobj_at)
+
+lemma transfer_caps_loop_valid_arch[Ipc_AI_2_assms]:
+  "\<And>slots caps ep buffer n mi.
+    \<lbrace>valid_arch_state and valid_objs and valid_mdb and K (distinct slots)
+         and (\<lambda>s. \<forall>x \<in> set slots. real_cte_at x s \<and> cte_wp_at (\<lambda>cap. cap = cap.NullCap) x s)
+         and transfer_caps_srcs caps\<rbrace>
+      transfer_caps_loop ep buffer n caps slots mi
+    \<lbrace>\<lambda>_. valid_arch_state\<rbrace>"
```

Comment:
Looks like I missed the indentation here in my review of the X64 commit, if it's not too much of a pain might as well fix it as well.

Reply:
already did it